alns/ALNS.py

```diff
@@ -246,18 +259,41 @@ def _eval_cand(
             A tuple of the best and current solution, along with the weight
             index.
         """
-        w_idx = _REJECT
+        outcome = self._determine_outcome(accept, best, curr, cand)
+        func = self._on_outcome.get(outcome)
+
+        if callable(func):
+            cand = func(cand, self._rnd_state, **kwargs)
+            outcome = self._determine_outcome(accept, best, curr, cand)
```

Comment:
Issue: Whenever a callback function is used in this implementation, `accept` is called twice and hence also is updated twice. This is problematic as the thresholds decrease twice as fast, which is undesired behavior.

I propose the convention that the callbacks are applied to the _new_ `current` solution. 
- We evaluate the candidate as we did before.
- Based on the outcome, we apply the corresponding callback if available.
    - For best, better, accepted, the new current solution is the candidate solution. The callback is applied to this candidate solution.
    - For rejected, the new current solution is the old current solution. The callback is applied to the old current solution.
- Check (again) if the post-callback solution is a new best or not.

Reply:
I get the issue, but I'm not 100% sure about the proposed solution. Particularly, I dislike that `on_reject` would never see the rejected candidate. That defeats some of the purpose.

I will think a bit about alternatives.

Reply:
Yes this works for me!